ement-room.el

```diff
@@ -2800,41 +2803,44 @@ ROOM defaults to the value of `ement-room'."
   ;; Is there an existing Emacs function to do this?  I couldn't find one.
   ;; Yes, maybe: `goto-address-mode'.  TODO: Try goto-address-mode.
-  (let ((buffer (alist-get 'buffer (ement-room-local ement-room))))
-    (with-temp-buffer
-      (insert string)
-      (goto-char (point-min))
-      (cl-loop with c = 1
-               with url = nil
-               while (re-search-forward (rx bow "http" (optional "s") "://" (1+ (not space)))
-                                        nil 'noerror)
-               do (setf url (buffer-substring-no-properties (match-beginning 0) (match-end 0)))
-               (make-text-button (match-beginning 0) (match-end 0)
-                                 'mouse-face 'highlight
-                                 'face 'link
-                                 'help-echo url
-                                 'action #'browse-url-at-mouse
-                                 'follow-link t)
-               if (and ement-room-url-previews
-                       (not (map-nested-elt (ement-event-content message) '(m.relates_to m.in_reply_to)))
-                       ;; spec recommends not getting previews for encrypted rooms
-                       (not (with-current-buffer buffer
-                              (cl-find-if (apply-partially #'equal "m.room.encryption")
-                                          (ement-room-invite-state ement-room)
-                                          :key #'ement-event-type))))
-               do (with-current-buffer buffer
-                    (ement-room--url-preview message url c))
-               (setf c (1+ c)))
-      (buffer-string))))
-
-(defun ement-room--url-preview (message link num)
-  "Fetch and insert a preview of LINK from MESSAGE. NUM is the index for this link."
-  (let* ((id (ement-event-id message))
+  (with-temp-buffer
+    (insert string)
+    (goto-char (point-min))
+    (cl-loop while (re-search-forward (rx bow "http" (optional "s") "://" (1+ (not space)))
+                                      nil 'noerror)
+             do (make-text-button (match-beginning 0) (match-end 0)
+                                  'mouse-face 'highlight
+                                  'face 'link
+                                  'help-echo (match-string 0)
+                                  'action #'browse-url-at-mouse
+                                  'follow-link t
+                                  'ement-url t))
+    (buffer-string)))
+
+(defun ement-room--preview-urls (event rendered)
+  "Parse RENDERED body of EVENT for URLs and fetch previews for them."
+  (if (and (not (alist-get 'url-previews (ement-event-local event)))
+           rendered)
+      (let ((room-buffer (alist-get 'buffer (ement-room-local ement-room))))
+        (with-temp-buffer
+          (insert rendered)
+          (goto-char (point-min))
+          (cl-loop with match = nil
+                   while (setf match (or (text-property-search-forward 'ement-url)
```

Comment:
Why not just do a regexp search for URLs on the message's plain-text body?  It will probably be faster, because Emacs's regexp engine is very well optimized, and the code will be much simpler.